py/torch_tensorrt/dynamo/conversion/converter_utils.py

```diff
@@ -295,22 +298,119 @@ def get_trt_tensor(
     Returns:
         A TensorRT ITensor that represents the given value.
     """
-    # TRT can not add constant for bool type. We do a work around to 1) cast it to int and 2)cast to bool later
-    # This is useful for logical operations which require input to be bool type
-    if isinstance(input_val, bool):
-        input_val = int(input_val)
-    elif isinstance(input_val, torch.Tensor) and (
-        input_val.dtype == torch.bool or input_val.dtype == torch.int64
+    # If the input is 64-bit, cast it to 32-bit for TRT freezing
+    if (
+        isinstance(input_val, torch.Tensor)
+        and ctx.compilation_settings.truncate_long_and_double
     ):
-        input_val = input_val.to(torch.int32)
-    elif isinstance(input_val, np.ndarray) and (
-        input_val.dtype == np.bool_ or input_val.dtype == np.int64
+        if input_val.dtype == torch.int64:
+            input_val = input_val.to(torch.int32)
+        elif input_val.dtype == torch.float64:
+            input_val = input_val.to(torch.float32)
+    elif (
+        isinstance(input_val, np.ndarray)
+        and ctx.compilation_settings.truncate_long_and_double
     ):
-        input_val = input_val.astype(np.int32)
+        if input_val.dtype == np.int64:
+            input_val = input_val.astype(np.int32)
+        elif input_val.dtype == np.float64:
+            input_val = input_val.astype(np.float32)
 
-    if isinstance(input_val, (torch.Tensor, np.ndarray, int, float)):
-        return create_constant(network, input_val, name, dtype)
+    if isinstance(input_val, (torch.Tensor, np.ndarray, int, float, bool)):
+        return create_constant(ctx, input_val, name, dtype)
     elif isinstance(input_val, TRTTensor):
         return input_val
     else:
         raise AssertionError(f"Cannot convert {input_val} to TRT constant")
+
+
+def enforce_tensor_types(
```

Comment:
Where did we land on using this to enforce scalar types as well?

Reply:
I did not add it, since I wasn't fully clear on the specification. For instance, should a user be able to specify `3: (int, float)`? Additionally, if they specify `3: (int, np,ndarray)`, would it be the `enforce_tensor_types` responsibility to cast that numpy array to integer if it was a float type? The latter seems out-of-scope for this function.

Reply:
Alternatively, if the specification was `3: (int, float)`, and we see an `np.ndarray`, should we attempt a cast (`int(array)`)? Or should it throw an error?

Reply:
If promote is true then enforce_types should try to cast it, for 

> Alternatively, if the specification was 3: (int, float), and we see an np.ndarray, should we attempt a cast (int(array))? Or should it throw an error?

Here we just run `.item()` and if it fails we thrown an error. 

> I did not add it, since I wasn't fully clear on the specification. For instance, should a user be able to specify 3: (int, float)? Additionally, if they specify 3: (int, np.ndarray), would it be the enforce_tensor_types responsibility to cast that numpy array to integer if it was a float type? The latter seems out-of-scope for this function

Here yes, you should be able to say, I want either a float32 or an int32. If the converter is provided a type that is in the set no conversion should happen, otherwise if promote_types is true then it should try to convert to the most preferred (first) type I think.

We could probably do this in another PR to get converter folks this feature sooner, but we should probably rename this function `enforce_types`

Reply:
Added #2349 with the proposed new features. This may need some additional discussion because there is a key distinction between Schema-equivalent-types, Tensor-equivalent-types, and data types. It seems that all three definitions are coming together here.

The intent of this function was to translate between Tensor-equivalent types, without consideration for the data types within those tensors. For instance, a converter could specify that it cannot handle `np.ndarray`, but it can handle `TRTTensor`, with no regard for the actual data types within. One step further would be to allow Schema-equivalent-types. That is - translate between Tensor types as needed, and Scalar types as needed, but no modification of data types within the Tensors. Finally is data types, which would probably require a new function header/schema altogether, since it requires that we also check the data type within the tensor and cast appropriately.

For now, I can leave this PR as-is and work on the other feature separately, to unblock any converter implementations that might depend on it. What do you think?

Reply:
Yeah lets merge this and work towards a schema type level enforcement mechanism